File: tests/test_balanced_group_system.py
import pytest
from balanced_group_system import BalancedGroupSystem

@pytest.fixture
def balanced_group_system():
  return BalancedGroupSystem(['Alice', 'Bob', 'Charlie', 'David', 'Eve'])

def test_init(balanced_group_system):
  assert balanced_group_system.members == ['Alice', 'Bob', 'Charlie', 'David', 'Eve']
  assert balanced_group_system.familiarity_matrix == {
    frozenset(('Alice', 'Bob')) : 0,
    frozenset(('Alice', 'Charlie')) : 0,
    frozenset(('Alice', 'David')) : 0,
    frozenset(('Alice', 'Eve')) : 0,
    frozenset(('Bob', 'Charlie')) : 0,
    frozenset(('Bob', 'David')) : 0,
    frozenset(('Bob', 'Eve')) : 0,
    frozenset(('Charlie', 'David')) : 0,
    frozenset(('Charlie', 'Eve')) : 0,
    frozenset(('David', 'Eve')) : 0,
  }

def test_add_member(balanced_group_system):
  balanced_group_system.add_member('Frank')
  assert balanced_group_system.members == ['Alice', 'Bob', 'Charlie', 'David', 'Eve', 'Frank']
  assert balanced_group_system.familiarity_matrix == {
    frozenset(('Alice', 'Bob')) : 0,
    frozenset(('Alice', 'Charlie')) : 0,
    frozenset(('Alice', 'David')) : 0,
    frozenset(('Alice', 'Eve')) : 0,
    frozenset(('Bob', 'Charlie')) : 0,
    frozenset(('Bob', 'David')) : 0,
    frozenset(('Bob', 'Eve')) : 0,
    frozenset(('Charlie', 'David')) : 0,
    frozenset(('Charlie', 'Eve')) : 0,
    frozenset(('David', 'Eve')) : 0,
    frozenset(('Frank', 'Alice')) : 0,
    frozenset(('Frank', 'Bob')) : 0,
    frozenset(('Frank', 'Charlie')) : 0,
    frozenset(('Frank', 'David')) : 0,
    frozenset(('Frank', 'Eve')) : 0,
  }

def test_remove_member(balanced_group_system):
  balanced_group_system.remove_member('Bob')
  assert balanced_group_system.members == ['Alice', 'Charlie', 'David', 'Eve']
  assert balanced_group_system.familiarity_matrix == {
    frozenset(('Alice', 'Charlie')) : 0,
    frozenset(('Alice', 'David')) : 0,
    frozenset(('Alice', 'Eve')) : 0,
    frozenset(('Charlie', 'David')) : 0,
    frozenset(('Charlie', 'Eve')) : 0,
    frozenset(('David', 'Eve')) : 0,
  }

def test_create_groups(balanced_group_system):
  groups = [['Alice', 'Bob'], ['Charlie', 'David', 'Eve']]
  balanced_group_system.create_groups(groups)
  assert balanced_group_system.group_history[0] == groups
  assert balanced_group_system.familiarity_matrix[frozenset(('Alice', 'Bob'))] == 2
  assert balanced_group_system.familiarity_matrix[frozenset(('Charlie', 'David'))] == 2
  assert balanced_group_system.familiarity_matrix[frozenset(('Charlie', 'Eve'))] == 2
  assert balanced_group_system.familiarity_matrix[frozenset(('David', 'Eve'))] == 2

def test_update_familiarity(balanced_group_system):
  group = ['Alice', 'Bob', 'Charlie']
  balanced_group_system.update_familiarity(group)
  assert balanced_group_system.familiarity_matrix[frozenset(('Alice', 'Bob'))] == 2
  assert balanced_group_system.familiarity_matrix[frozenset(('Alice', 'Charlie'))] == 2
  assert balanced_group_system.familiarity_matrix[frozenset(('Bob', 'Charlie'))] == 2

def test_calculate_balanced_groups(balanced_group_system):
  groups = balanced_group_system.calculate_balanced_groups(2, balanced_group_system.members)
  assert len(groups) == 2
  assert all(len(group) >= 2 for group in groups)

def test_create_balanced_groups(balanced_group_system):
  groups = balanced_group_system.create_balanced_groups(2)
  assert len(groups) == 2
  assert all(len(group) >= 2 for group in groups)

File: balanced_group_system/test_balanced_group_system.cpp
#include "balanced_group_system.hpp"
#include <iostream>
#include <vector>
#include <cassert>
#include <stdexcept>

void test_init() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    BalancedGroupSystem bgs(members);
    assert(bgs.members == members);
    assert(bgs.familiarity_matrix == std::vector<std::vector<int>>(5, std::vector<int>(5,0)));
}

void test_add_member() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    BalancedGroupSystem bgs(members);
    bgs.add_member("Frank");
    assert(bgs.members.size() == 6);
    assert(bgs.familiarity_matrix.size() == 6);
    assert(bgs.familiarity_matrix == std::vector<std::vector<int>>(6, std::vector<int>(6,0)));
}

void test_remove_member() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    BalancedGroupSystem bgs(members);
    bgs.remove_member("Bob");
    assert(bgs.members.size() == 4);
    assert(bgs.familiarity_matrix.size() == 4);
    assert(bgs.familiarity_matrix == std::vector<std::vector<int>>(4, std::vector<int>(4,0)));
}

void test_create_groups() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    BalancedGroupSystem bgs(members);
    std::vector<std::vector<std::string>> groups = {
        {"Alice", "Bob"},
        {"Charlie", "David", "Eve"}
    };
    bgs.create_groups(groups);
    assert(bgs.familiarity_matrix[0][1] == 1);
    assert(bgs.familiarity_matrix[2][3] == 1);
    assert(bgs.familiarity_matrix[2][4] == 1);
    assert(bgs.familiarity_matrix[3][4] == 1);
}

void test_update_familiarity() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    BalancedGroupSystem bgs(members);
    std::vector<std::string> group = {"Alice", "Bob", "Charlie"};
    bgs.update_familiarity(group);
    assert(bgs.familiarity_matrix[0][1] == 1);
    assert(bgs.familiarity_matrix[0][2] == 1);
    assert(bgs.familiarity_matrix[1][2] == 1);
}

void test_calculate_balanced_groups() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    BalancedGroupSystem bgs(members);
    std::vector<std::string> test_members;
    for(std::string member : members) {
      test_members.push_back(member);
    }
    std::vector<std::vector<std::string>> groups = bgs.calculate_balanced_groups(2, test_members);
    assert(groups.size() == 2);
    for(std::vector<std::string> group : groups) {
      assert(group.size() >= 2);
    }
}

void test_create_balanced_groups() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    BalancedGroupSystem bgs(members);
    std::vector<std::vector<std::string>> groups = bgs.create_balanced_groups(2);
    assert(groups.size() == 2);
    for(std::vector<std::string> group : groups) {
      assert(group.size() >= 2);
    }
}

int main(){
    test_init();
    test_add_member();
    test_remove_member();
    test_create_groups();
    test_update_familiarity();
    test_calculate_balanced_groups();
    test_create_balanced_groups();
}
